### PR TITLE
[1.19.3] Fix Datagen Tests and Providers

### PR DESCRIPTION
**Pull request by ChampionAsh5357**
This PR fixes the following things related to data generations

- Removes the abstract keyword from the `ForgeAdvancementProvider` since it should not be subclassed normally and adjusts the datagen tests to accommodate the changes.
- Fixes some errors from missed changes from the `PackOutput`.
- Removes `JsonCodecProvider#forDatapackRegistry` in favor of using `RegistriesDatapackGenerator` from vanilla
- Adds the ability to dump dynamic registry objects by mod id in `RegistriesDatapackGenerator`
- Fix inlining of registry objects when using `DatapackBuiltinEntriesProvider`
- Fix the `BiomeModifierTest` so that we can test generation with the new provider

### CODE REVIEW

File: patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
 +++ b/net/minecraft/data/registries/RegistriesDatapackGenerator.java
-@@ -31,7 +_,7 @@
+@@ -22,16 +_,28 @@
+    private static final Logger f_254654_ = LogUtils.getLogger();
+    private final PackOutput f_254743_;
+    private final CompletableFuture<HolderLookup.Provider> f_254747_;
++   private final java.util.function.Predicate<String> namespacePredicate;
+ 
++   /**
++    * @deprecated Use mod-specific version instead
```
Comment (sciwhiz12) [inline patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch]: This should have the `Forge: ` prefix.
```suggestion
+    * @deprecated Forge: Use mod-specific version instead
```

File: patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch [hunk 2]
```diff
@@ -1,6 +1,27 @@
...
++    */
++   @Deprecated
+    public RegistriesDatapackGenerator(PackOutput p_256643_, CompletableFuture<HolderLookup.Provider> p_255780_) {
+       this.f_254747_ = p_255780_;
+       this.f_254743_ = p_256643_;
++      this.namespacePredicate = namespace -> true;
++   }
++
++   public RegistriesDatapackGenerator(PackOutput p_256643_, CompletableFuture<HolderLookup.Provider> p_255780_, java.util.Set<String> modIds) {
```
Comment (sciwhiz12) [inline patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch]: This should have an EXC entry.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch [hunk 3]
```diff
@@ -1,6 +1,32 @@
...
++@Deprecated
+ public class RegistriesDatapackGenerator implements DataProvider {
+    private static final Logger f_254654_ = LogUtils.getLogger();
+    private final PackOutput f_254743_;
+    private final CompletableFuture<HolderLookup.Provider> f_254747_;
++   private final java.util.function.Predicate<String> namespacePredicate;
+ 
++   @Deprecated
+    public RegistriesDatapackGenerator(PackOutput p_256643_, CompletableFuture<HolderLookup.Provider> p_255780_) {
+       this.f_254747_ = p_255780_;
```
Comment (Matyrobbrt) [inline patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch]: This constructor should instead delegate to the other one.

File: src/main/java/net/minecraftforge/common/data/DatapackBuiltinEntriesProvider.java [hunk 4]
```diff
@@ -5,29 +5,63 @@
...
+ * An extension of the {@link RegistriesDatapackGenerator} which properly handles
+ * referencing existing dynamic registry objects within another dynamic registry
+ * object.
+ */
 public class DatapackBuiltinEntriesProvider extends RegistriesDatapackGenerator
 {
-    public DatapackBuiltinEntriesProvider(PackOutput output, Supplier<HolderLookup.Provider> registriesSupplier)
+
+    /**
+     * Default constructor.
```
Comment (Matyrobbrt) [inline src/main/java/net/minecraftforge/common/data/DatapackBuiltinEntriesProvider.java]: Useless Javadoc, please include useful information.
Same for the one below.

File: src/test/java/net/minecraftforge/debug/world/BiomeModifierTest.java [hunk 5]
```diff
@@ -106,81 +135,39 @@ public BiomeModifierTest()
...
-                REMOVE_FOREST_SKELETONS_RL, removeSpawn)));
+        event.getGenerator().addProvider(event.includeServer(), (DataProvider.Factory<BiomeModifiers>) output -> new BiomeModifiers(output, event.getLookupProvider()));
     }
 
-    public record TestModifier(HolderSet<Biome> biomes, Precipitation precipitation, int waterColor) implements BiomeModifier
+    private static class BiomeModifiers extends DatapackBuiltinEntriesProvider
     {
-        private static final RegistryObject<Codec<? extends BiomeModifier>> SERIALIZER = RegistryObject.create(MODIFY_BIOMES_RL, ForgeRegistries.Keys.BIOME_MODIFIER_SERIALIZERS, MODID);
+
+        public BiomeModifiers(PackOutput output, CompletableFuture<HolderLookup.Provider> registries) {
```
Comment (Matyrobbrt) [inline src/test/java/net/minecraftforge/debug/world/BiomeModifierTest.java]: Brackets on new line.

File: src/test/java/net/minecraftforge/debug/DataGeneratorTest.java [hunk 6]
```diff
@@ -1014,54 +1024,45 @@ public String getName() {
...
 
-    public static class Advancements extends ForgeAdvancementProvider
-    {
-
-
-        public Advancements(PackOutput output, CompletableFuture<HolderLookup.Provider> providerCompletableFuture, ExistingFileHelper existingFileHelper)
-        {
-            super(output, providerCompletableFuture, existingFileHelper, List.of(new Provider()));
-        }
+    private static class Advancements implements ForgeAdvancementProvider.AdvancementGenerator {
```
Comment (Matyrobbrt) [inline src/test/java/net/minecraftforge/debug/DataGeneratorTest.java]: Brackets on new line.

File: src/main/resources/META-INF/accesstransformer.cfg [hunk 7]
```diff
@@ -452,4 +453,4 @@ public net.minecraft.world.level.levelgen.feature.trunkplacers.TrunkPlacerType <
 protected net.minecraft.world.level.portal.PortalForcer f_77648_ # level
 public net.minecraft.world.level.storage.LevelResource <init>(Ljava/lang/String;)V # constructor
 private-f net.minecraft.world.level.storage.loot.LootPool f_79028_ # rolls
-private-f net.minecraft.world.level.storage.loot.LootPool f_79029_ # bonusRolls
```
Comment (Matyrobbrt) [inline src/main/resources/META-INF/accesstransformer.cfg:455]: EOL whitespace change.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch (hunk 3 above)
Comment (ChampionAsh5357) [reply patches/minecraft/net/minecraft/data/registries/RegistriesDatapackGenerator.java.patch]: The constructor takes in set of mod ids while the predicate always returns true. Are you suggesting I pass in all the mod ids or return null to force the set in that case?

**Review by Matyrobbrt — APPROVED** [review]

**Pull request merged**